### PR TITLE
WIP: Initial working captcha implementation

### PR DESCRIPTION
**Pull request by sinukaarel**
First implementation of original Magento capcha.

image

The idea is to display captcha inside newsletter form conditionally when Smaily module is enabled.

One implementation would be to override the whole newsletter signup form and add captcha section to the template directly. The second option that I chose was to go with Magento way and extend original/sore owners current template and add captcha section to it programmatically. 

This captcha section is only shown/added when the formID `smaily_captcha` is added to  `etc/config.xml` file form section.

Currently it is forced to do show as it is hardcoded to config section.
There are now couple of options how to handle this problem.
- Create a option in the plugins **Newsletter Subscription Form** section that allows customer to enable Magento capcha in Newsletter form. (No idea how to do this yet, but i think that could be done)
- Obmit the hardcode part as customer can enable captcha also from `Stores->Customers->Customer Configuration->Captcha->Forms` section. Show only note/comment in Smaily options page.
- Go as it is, when you enable module there is captcha. Downside is that maybe they have some captcha plugin installed and now we are overriding it.
- Don't show newsletter sign-up form at all when our module is installed and captcha is not enabled. Add note to settings page newsletter section.

### CODE REVIEW

**Comment by sinukaarel** [comment]
Implementation on google invisible reCAPTCHA
image

To continue on initial comment I went with few of the options above:
> * Create a option in the plugins **Newsletter Subscription Form** section that allows customer to enable Magento capcha in Newsletter form. (No idea how to do this yet, but i think that could be done)

- Customers can now chose if they want to enable Magento CAPTCHA or Google reCAPTCHA

image


> * Obmit the hardcode part as customer can enable captcha also from `Stores->Customers->Customer Configuration->Captcha->Forms` section. Show only note/comment in Smaily options page.

- Still added Smaily to the default forms to show CAPTCHA in. Otherwise I found that CAPTCHA block wont be rendered.

> * Don't show newsletter sign-up form at all when our module is installed and captcha is not enabled. Add note to settings page newsletter section.

-  When customer has **disabled Magento built in CAPTCHA** or has changed the default behavior then **newsletter form wont be displayed** if Magento CAPTCHA is selected for bot control. Added note about this in settings page also.

image

File: view/frontend/templates/subscribe.phtml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<?php
+//Don't care if form collection is disabled as bots would get added also with sync only.
```
Comment (kaittodesk) [inline view/frontend/templates/subscribe.phtml]: Formatting: missing space between `//` and `Don't`.

File: Block/Subscribe.php [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    public function getOriginalTemplate()
+    {
+        return $this->
+                getLayout()->
+                createBlock("Magento\Newsletter\Block\Subscribe")->
+                setTemplate('Magento_Newsletter::subscribe.phtml');
+    }
+
+    /**
+     * Get newsletter template with capcha section.
```
[the remark this answers is not in the document]
Comment (sinukaarel) [reply Block/Subscribe.php]: Fixed in 5b01525b67a526cd7dd2ee22e2451f31d2d632ce

File: Block/Subscribe.php [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            $captchaTemplate = $this->getBlockHtml('smaily.recaptcha');
+        } else {
+            // May return empty string if built-in captcha is disabled.
+            $captchaTemplate = $this->getBlockHtml('smaily.captcha');
+        }
+
+        // Only show newsletter form when captcha and collection is enabled.
+        if ($captchaTemplate && $subscribeCollectionEnabled) {
+            if ($captchaType === 'magento_captcha') {
+                // Remove newsletter class (keep only block class) from original form as it messes up css.
```
[the remark this answers is not in the document]
Comment (sinukaarel) [reply Block/Subscribe.php]: fixed in 5b01525b67a526cd7dd2ee22e2451f31d2d632ce

File: Helper/Data.php [hunk 4]
```diff
@@ -145,6 +141,36 @@ public function getGeneralConfig($code, $storeId = null)
...
+     *
+     * @return string Captcha type.
+     */
+    public function getCaptchaType()
+    {
+        return $this->getGeneralConfig('captcha_type');
+    }
+
+    /**
+     * Get reCAHPTCHA public api key.
```
[the remark this answers is not in the document]
Comment (sinukaarel) [reply Helper/Data.php]: Fixed in 5b01525b67a526cd7dd2ee22e2451f31d2d632ce

File: Helper/Data.php [hunk 5]
```diff
@@ -145,6 +141,36 @@ public function getGeneralConfig($code, $storeId = null)
...
+     *
+     * @return string public key.
+     */
+    public function getCaptchaApiKey()
+    {
+        return $this->getGeneralConfig('captcha_api_key');
+    }
+
+    /**
+     * Get reCAPTCHA private api key.
```
[the remark this answers is not in the document]
Comment (sinukaarel) [reply Helper/Data.php]: Fixed in 5b01525b67a526cd7dd2ee22e2451f31d2d632ce

File: Helper/Data.php [hunk 6]
```diff
@@ -482,6 +508,36 @@ public function validateApiCredentrials($subdomain, $username, $password)
         return $response;
     }
 
+    /**
+     * Validate reCaptcha response value.
```
[the remark this answers is not in the document]
Comment (sinukaarel) [reply Helper/Data.php]: Fixed in 5b01525b67a526cd7dd2ee22e2451f31d2d632ce

File: view/frontend/templates/subscribe.phtml (hunk 1 above)
Comment (sinukaarel) [reply view/frontend/templates/subscribe.phtml]: Fixed in 5b01525b67a526cd7dd2ee22e2451f31d2d632ce

File: Block/Subscribe.php [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            }
+             // Select form and action section.
+            $form = $originalDOM->getElementsByTagName('form')->item(0);
+            $actionsSection = $xPath->query('//div[@class="actions"]')->item(0);
+
+            // Add capcha section before action section.
+            $captcha = $originalDOM->createDocumentFragment();
+            $captcha->appendXML($captchaTemplate);
+            $form->insertBefore($captcha, $actionsSection);
+            return $originalDOM->saveHTML();
```
[the remark this answers is not in the document]
Comment (sinukaarel) [reply Block/Subscribe.php]: Yes, template got extra tags that won't be necessary in this output. Removed those and also got a good read about bitwise OR. fb9e5b1252b124b31f51d17456c631af6acc94a5

File: Observer/NewsletterCaptchaObserver.php [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+            $captcha = $this->helper->getCaptcha($formId);
+            if (!$captcha->isCorrect($this->captchaStringResolver->resolve($controller->getRequest(), $formId))) {
+                $this->messageManager->addError(__('Incorrect CAPTCHA.'));
+                $this->getDataPersistor()->set($formId, $controller->getRequest()->getPostValue());
+                $this->actionFlag->set('', \Magento\Framework\App\Action\Action::FLAG_NO_DISPATCH, true);
+                $this->redirect->redirect($controller->getResponse(), $this->redirect->getRedirectUrl());
+            }
+        }
+
+        if ($captchaType === 'google_captcha') {
```
[the remark this answers is not in the document]
Comment (sinukaarel) [reply Observer/NewsletterCaptchaObserver.php]: Done fb9e5b1252b124b31f51d17456c631af6acc94a5

**Pull request merged**